### ISSUE TITLE
Non-negative matrix factorization using Tensorflow

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#4092]
Non-negative matrix factorization using Tensorflow<br>
<br>
http://bit.ly/2JzoOOy<br>
<br>
About 0 Minutes<br>
<br>
<div>
<div>
<div>
<div>
<p>In a <a href="https://nipunbatra.github.io/blog/2017/nmf-nnls.html">previous post</a>, we had seen how to perfom non-negative matrix factorization (NNMF) using non-negative least squares (NNLS). In this post, we will look at performing NNMF using TensorFlow. As before, we will look at factorizing matrices that may contain missing entries (for the problem of movie recommendation, etc.). Like explained in a <a href="https://nipunbatra.github.io/blog/2017/nn-linear-regression-tensorflow.html">previous post</a>, we would be using projected gradient descent for this problem. In other words, we would be computing the gradient, then ensuring that the weights are non-negative, then perform gradient descent.</p>
</div>
</div>
<div>
<div>
<p>In&nbsp;[1]:</p>
<div>
<div>
<div>
<pre>
<span>import</span> <span>tensorflow</span> <span>as</span> <span>tf</span>
<span>import</span> <span>numpy</span> <span>as</span> <span>np</span>
<span>import</span> <span>pandas</span> <span>as</span> <span>pd</span>
<span>np</span><span>.</span><span>random</span><span>.</span><span>seed</span><span>(</span><span>0</span><span>)</span>
</pre>
</div>
</div>
</div>
</div>
</div>
<div>
<div>
<div>
<h3>Creating the matrix to be decomposed<a href="#Creating-the-matrix-to-be-decomposed">&para;</a>
</h3>
</div>
</div>
</div>
<div>
<div>
<p>In&nbsp;[2]:</p>
<div>
<div>
<div>
<pre>
<span>A_orig</span> <span>=</span> <span>np</span><span>.</span><span>array</span><span>([[</span><span>3</span><span>,</span> <span>4</span><span>,</span> <span>5</span><span>,</span> <span>2</span><span>],</span>
                   <span>[</span><span>4</span><span>,</span> <span>4</span><span>,</span> <span>3</span><span>,</span> <span>3</span><span>],</span>
                   <span>[</span><span>5</span><span>,</span> <span>5</span><span>,</span> <span>4</span><span>,</span> <span>4</span><span>]],</span> <span>dtype</span><span>=</span><span>np</span><span>.</span><span>float32</span><span>)</span><span>.</span><span>T</span>

<span>A_orig_df</span> <span>=</span> <span>pd</span><span>.</span><span>DataFrame</span><span>(</span><span>A_orig</span><span>)</span>
</pre>
</div>
</div>
</div>
</div>
</div>
<div>
<div>
<p>In&nbsp;[3]:</p>
<div>
<div>
<div>
<pre>
<span>A_orig_df</span> <span>#(4 users, 3 movies)</span>
</pre>
</div>
</div>
</div>
</div>
<div>
<div>
<div>
<p>Out[3]:</p>
<div>
<div>
<table border="1">
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
</tr>
</thead>
<tbody>
<tr>
<th>0</th>
<td>3.0</td>
<td>4.0</td>
<td>5.0</td>
</tr>
<tr>
<th>1</th>
<td>4.0</td>
<td>4.0</td>
<td>5.0</td>
</tr>
<tr>
<th>2</th>
<td>5.0</td>
<td>3.0</td>
<td>4.0</td>
</tr>
<tr>
<th>3</th>
<td>2.0</td>
<td>3.0</td>
<td>4.0</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
</div>
</div>
<div>
<div>
<p>In&nbsp;[4]:</p>
<div>
<div>
<div>
<pre>
<span>A_df_masked</span> <span>=</span> <span>A_orig_df</span><span>.</span><span>copy</span><span>()</span>
<span>A_df_masked</span><span>.</span><span>iloc</span><span>[</span><span>0</span><span>,</span><span>0</span><span>]</span><span>=</span><span>np</span><span>.</span><span>NAN</span>
</pre>
</div>
</div>
</div>
</div>
</div>
<div>
<div>
<p>In&nbsp;[5]:</p>
<div>
<div>
<div>
<pre>
<span>np_mask</span> <span>=</span> <span>A_df_masked</span><span>.</span><span>notnull</span><span>()</span>
<span>np_mask</span>
</pre>
</div>
</div>
</div>
</div>
<div>
<div>
<div>
<p>Out[5]:</p>
<div>
<div>
<table border="1">
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
</tr>
</thead>
<tbody>
<tr>
<th>0</th>
<td>False</td>
<td>True</td>
<td>True</td>
</tr>
<tr>
<th>1</th>
<td>True</td>
<td>True</td>
<td>True</td>
</tr>
<tr>
<th>2</th>
<td>True</td>
<td>True</td>
<td>True</td>
</tr>
<tr>
<th>3</th>
<td>True</td>
<td>True</td>
<td>True</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
</div>
</div>
<div>
<div>
<p>In&nbsp;[6]:</p>
<div>
<div>
<div>
<pre>
<span># Boolean mask for computing cost only on valid (not missing) entries</span>
<span>tf_mask</span> <span>=</span> <span>tf</span><span>.</span><span>Variable</span><span>(</span><span>np_mask</span><span>.</span><span>values</span><span>)</span>

<span>A</span> <span>=</span> <span>tf</span><span>.</span><span>constant</span><span>(</span><span>A_df_masked</span><span>.</span><span>values</span><span>)</span>
<span>shape</span> <span>=</span> <span>A_df_masked</span><span>.</span><span>values</span><span>.</span><span>shape</span>

<span>#latent factors</span>
<span>rank</span> <span>=</span> <span>3</span> 

<span># Initializing random H and W</span>
<span>temp_H</span> <span>=</span> <span>np</span><span>.</span><span>random</span><span>.</span><span>randn</span><span>(</span><span>rank</span><span>,</span> <span>shape</span><span>[</span><span>1</span><span>])</span><span>.</span><span>astype</span><span>(</span><span>np</span><span>.</span><span>float32</span><span>)</span>
<span>temp_H</span> <span>=</span> <span>np</span><span>.</span><span>divide</span><span>(</span><span>temp_H</span><span>,</span> <span>temp_H</span><span>.</span><span>max</span><span>())</span>

<span>temp_W</span> <span>=</span> <span>np</span><span>.</span><span>random</span><span>.</span><span>randn</span><span>(</span><span>shape</span><span>[</span><span>0</span><span>],</span> <span>rank</span><span>)</span><span>.</span><span>astype</span><span>(</span><span>np</span><span>.</span><span>float32</span><span>)</span>
<span>temp_W</span> <span>=</span> <span>np</span><span>.</span><span>divide</span><span>(</span><span>temp_W</span><span>,</span> <span>temp_W</span><span>.</span><span>max</span><span>())</span>

<span>H</span> <span>=</span>  <span>tf</span><span>.</span><span>Variable</span><span>(</span><span>temp_H</span><span>)</span>
<span>W</span> <span>=</span> <span>tf</span><span>.</span><span>Variable</span><span>(</span><span>temp_W</span><span>)</span>
<span>WH</span> <span>=</span> <span>tf</span><span>.</span><span>matmul</span><span>(</span><span>W</span><span>,</span> <span>H</span><span>)</span>
</pre>
</div>
</div>
</div>
</div>
</div>
<div>
<div>
<p>In&nbsp;[7]:</p>
<div>
<div>
<div>
<pre>
<span>#cost of Frobenius norm</span>
<span>cost</span> <span>=</span> <span>tf</span><span>.</span><span>reduce_sum</span><span>(</span><span>tf</span><span>.</span><span>pow</span><span>(</span><span>tf</span><span>.</span><span>boolean_mask</span><span>(</span><span>A</span><span>,</span> <span>tf_mask</span><span>)</span> <span>-</span> <span>tf</span><span>.</span><span>boolean_mask</span><span>(</span><span>WH</span><span>,</span> <span>tf_mask</span><span>),</span> <span>2</span><span>))</span>
</pre>
</div>
</div>
</div>
</div>
</div>
<div>
<div>
<p>In&nbsp;[8]:</p>
<div>
<div>
<div>
<pre>
<span># Learning rate</span>
<span>lr</span> <span>=</span> <span>0.001</span>
<span># Number of steps</span>
<span>steps</span> <span>=</span> <span>1000</span>
<span>train_step</span> <span>=</span> <span>tf</span><span>.</span><span>train</span><span>.</span><span>GradientDescentOptimizer</span><span>(</span><span>lr</span><span>)</span><span>.</span><span>minimize</span><span>(</span><span>cost</span><span>)</span>
<span>init</span> <span>=</span> <span>tf</span><span>.</span><span>global_variables_initializer</span><span>()</span>
</pre>
</div>
</div>
</div>
</div>
</div>
<div>
<div>
<p>In&nbsp;[9]:</p>
<div>
<div>
<div>
<pre>
<span># Clipping operation. This ensure that W and H learnt are non-negative</span>
<span>clip_W</span> <span>=</span> <span>W</span><span>.</span><span>assign</span><span>(</span><span>tf</span><span>.</span><span>maximum</span><span>(</span><span>tf</span><span>.</span><span>zeros_like</span><span>(</span><span>W</span><span>),</span> <span>W</span><span>))</span>
<span>clip_H</span> <span>=</span> <span>H</span><span>.</span><span>assign</span><span>(</span><span>tf</span><span>.</span><span>maximum</span><span>(</span><span>tf</span><span>.</span><span>zeros_like</span><span>(</span><span>H</span><span>),</span> <span>H</span><span>))</span>
<span>clip</span> <span>=</span> <span>tf</span><span>.</span><span>group</span><span>(</span><span>clip_W</span><span>,</span> <span>clip_H</span><span>)</span>
</pre>
</div>
</div>
</div>
</div>
</div>
<div>
<div>
<p>In&nbsp;[10]:</p>
<div>
<div>
<div>
<pre>
<span>steps</span> <span>=</span> <span>1000</span>
<span>with</span> <span>tf</span><span>.</span><span>Session</span><span>()</span> <span>as</span> <span>sess</span><span>:</span>
    <span>sess</span><span>.</span><span>run</span><span>(</span><span>init</span><span>)</span>
    <span>for</span> <span>i</span> <span>in</span> <span>range</span><span>(</span><span>steps</span><span>):</span>
        <span>sess</span><span>.</span><span>run</span><span>(</span><span>train_step</span><span>)</span>
        <span>sess</span><span>.</span><span>run</span><span>(</span><span>clip</span><span>)</span>
        <span>if</span> <span>i</span><span>%</span><span>100</span>==0:
            <span>print</span><span>(</span><span>"</span><span>\n</span><span>Cost: </span><span>%f</span><span>"</span> <span>%</span> <span>sess</span><span>.</span><span>run</span><span>(</span><span>cost</span><span>))</span>
            <span>print</span><span>(</span><span>"*"</span><span>*</span><span>40</span><span>)</span>
    <span>learnt_W</span> <span>=</span> <span>sess</span><span>.</span><span>run</span><span>(</span><span>W</span><span>)</span>
    <span>learnt_H</span> <span>=</span> <span>sess</span><span>.</span><span>run</span><span>(</span><span>H</span><span>)</span>
</pre>
</div>
</div>
</div>
</div>
<div>
<div>
<div>
<div>
<pre>
Cost: 148.859848
****************************************

Cost: 3.930172
****************************************

Cost: 2.068570
****************************************

Cost: 1.418309
****************************************

Cost: 0.819721
****************************************

Cost: 0.399933
****************************************

Cost: 0.176080
****************************************

Cost: 0.079007
****************************************

Cost: 0.041353
****************************************

Cost: 0.027041
****************************************
</pre>
</div>
</div>
</div>
</div>
</div>
<div>
<div>
<div>
<h3>Computing the prediction<a href="#Computing-the-prediction">&para;</a>
</h3>
</div>
</div>
</div>
<div>
<div>
<div>
<div>
<p>Out[11]:</p>
<div>
<pre>
array([[ 0.86129224,  1.3388027 ,  1.97224879],
       [ 2.16338873,  0.97277433,  1.17212451],
       [ 0.25879648,  1.07861733,  1.09541821]], dtype=float32)
</pre>
</div>
</div>
</div>
</div>
</div>
<div>
<div>
<div>
<div>
<p>Out[12]:</p>
<div>
<pre>
array([[ 1.15797794,  0.97454673,  1.41825044],
       [ 1.44136858,  1.16967547,  0.79135358],
       [ 0.81640321,  1.98227394,  0.02636297],
       [ 1.38819814,  0.29285902,  0.8031919 ]], dtype=float32)
</pre>
</div>
</div>
</div>
</div>
</div>
<div>
<div>
<p>In&nbsp;[13]:</p>
<div>
<div>
<div>
<pre>
<span>pred</span> <span>=</span> <span>np</span><span>.</span><span>dot</span><span>(</span><span>learnt_W</span><span>,</span> <span>learnt_H</span><span>)</span>
<span>pred_df</span> <span>=</span> <span>pd</span><span>.</span><span>DataFrame</span><span>(</span><span>pred</span><span>)</span>
<span>pred_df</span><span>.</span><span>round</span><span>()</span>
</pre>
</div>
</div>
</div>
</div>
<div>
<div>
<div>
<p>Out[13]:</p>
<div>
<div>
<table border="1">
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
</tr>
</thead>
<tbody>
<tr>
<th>0</th>
<td>3.0</td>
<td>4.0</td>
<td>5.0</td>
</tr>
<tr>
<th>1</th>
<td>4.0</td>
<td>4.0</td>
<td>5.0</td>
</tr>
<tr>
<th>2</th>
<td>5.0</td>
<td>3.0</td>
<td>4.0</td>
</tr>
<tr>
<th>3</th>
<td>2.0</td>
<td>3.0</td>
<td>4.0</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
</div>
</div>
<div>
<div>
<p>Not bad! Just to recall our originial matrix.</p>
</div>
</div>
<div>
<div>
<div>
<div>
<p>Out[14]:</p>
<div>
<div>
<table border="1">
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
</tr>
</thead>
<tbody>
<tr>
<th>0</th>
<td>3.0</td>
<td>4.0</td>
<td>5.0</td>
</tr>
<tr>
<th>1</th>
<td>4.0</td>
<td>4.0</td>
<td>5.0</td>
</tr>
<tr>
<th>2</th>
<td>5.0</td>
<td>3.0</td>
<td>4.0</td>
</tr>
<tr>
<th>3</th>
<td>2.0</td>
<td>3.0</td>
<td>4.0</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via nipunbatra.github.io http://bit.ly/2Eq3lnf<br>
<br>
May 23, 2019 at 02:43PM
